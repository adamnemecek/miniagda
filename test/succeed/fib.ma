-- 2012-01-22 parameters gone from constructors

data Nat : Set {
  zero : Nat;
  succ : (n : Nat) -> Nat
}

fun add : Nat -> Nat -> Nat {
  add zero = \y -> y;
  add (succ x) = \y -> succ (add x y)
}

sized codata Stream : Size -> Set {
  cons : (i : Size) -> Nat -> Stream i -> Stream ($ i)
}

fun tail : Stream # -> Stream # {
  tail (cons .# x xs) = xs
}

fun head : Stream # -> Nat {
  head (cons .# x xs) = x
}

{-
norec head : (i : Size) -> Stream ($ i) -> Nat {
  head .($ i) (cons i n ns) = n
}

cofun zipWith :  (Nat -> Nat -> Nat ) -> ( i : Size )
                -> Stream i -> Stream i -> Stream i {
  zipWith f ($ i) as bs =
        cons i (f (head i as) (head i bs))  (zipWith f i (tail i as) (tail i bs))
}
-}

fun nth : Nat -> Stream # -> Nat {
  nth zero xs = head xs;
  nth (succ x) xs = nth x (tail xs)
}

let one : Nat = (succ zero)

cofun fib' : (x : Nat ) -> (y : Nat ) -> (i : Size ) -> Stream i
{
  fib' x y ($ i) = cons _ x (fib' y (add x y) _)
}
let fib : Stream # = (fib' one one _)


let four : Nat = (succ (succ (succ one)))

-- fib(four) = 5
eval let fibfour : Nat = nth four fib


--------------------------------------------
--------------------------------------------

data Leq : Nat -> Nat -> Set
{
lqz : (x : Nat ) -> Leq zero x ;
lqs : (x : Nat ) -> (y : Nat ) -> Leq x y -> Leq (succ x) (succ y)
}

sized codata Increasing : Size -> Stream # -> Set
{
inc : (i : Size ) -> (x : Nat ) -> (y : Nat ) -> Leq x y -> (tl : Stream # ) ->
      Increasing i (cons # y tl) ->
      Increasing ($ i) (cons # x (cons # y tl))
}


data Eq (+ A : Set)(a : A) : A -> Set
{ refl : Eq A a a
}

let proof : Eq (Stream #) (tail fib) (tail fib) = refl



let double : Stream # -> Stream # = \s -> cons _ (head s) s

data Bool : Set
{
tt : Bool;
ff : Bool
}

fun leq : Nat -> Nat -> Bool
{
leq zero y = tt;
leq (succ x) zero = ff ;
leq (succ x) (succ y) = leq x y
}

fun ite : Bool -> (A : Set ) -> A -> A -> A
{
ite tt A a1 a2 = a1;
ite ff A a1 a2 = a2
}

cofun merge : (i : Size ) -> (Nat -> Nat -> Bool) -> Stream # -> Stream # -> Stream i
{
merge ($ i) le (cons .# x xs) (cons .# y ys) =
      ite (le x y) (Stream _)
         (cons _ x (merge _ le xs (cons _ y ys)))
         (cons _ y (merge _ le (cons _ x xs) ys))
}

fun first : (A : Set ) -> (B : Set ) -> A -> B -> A
{
first A B a b = a
}

--------------------

cofun map : (i : Size) -> (Nat -> Nat) -> Stream i -> Stream i
{
map ($ i) f (cons .i x xl) = cons _ (f x) (map _ f xl)
}

{-
-- 2012-01-22 constructor are no longer inferable!
let suc : Nat -> Nat = \ x -> succ x
-- 2012-01-25 constructor recognition also for function types
-}

cofun evil : (i : Size) -> Stream i
{
evil ($ i) = map _ succ (cons _ zero (evil _))
}

-- eval const zzz : Nat = head # (z #)

