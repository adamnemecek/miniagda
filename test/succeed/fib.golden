--- opening "fib.ma" ---
--- scope checking ---
--- type checking ---
type  Nat : Set
term  Nat.zero : < Nat.zero : Nat >
term  Nat.succ : ^(n : Nat) -> < Nat.succ n : Nat >
term  add : Nat -> Nat -> Nat
{ add Nat.zero = \ y -> y
; add (Nat.succ x) = \ y -> Nat.succ (add x y)
}
type  Stream : - Size -> Set
term  Stream.cons : .[i : Size] -> ^(y1 : Nat) -> ^(y2 : Stream i) -> < Stream.cons i y1 y2 : Stream $i >
term  tail : Stream # -> Stream #
{ tail (Stream.cons [.#] x xs) = xs
}
term  head : Stream # -> Nat
{ head (Stream.cons [.#] x xs) = x
}
term  nth : Nat -> Stream # -> Nat
{ nth Nat.zero xs = head xs
; nth (Nat.succ x) xs = nth x (tail xs)
}
term  one : Nat
term  one = Nat.succ Nat.zero
term  fib' : (x : Nat) -> (y : Nat) -> .[i : Size] -> Stream i
{ fib' x y $[i < #] = Stream.cons [i] x (fib' y (add x y) [i])
}
term  fib : Stream #
term  fib = fib' one one [?2]
term  four : Nat
term  four = Nat.succ (Nat.succ (Nat.succ one))
term  fibfour : Nat
term  fibfour = nth four fib
type  Leq : ^ Nat -> ^ Nat -> Set
term  Leq.lqz : .[x : Nat] -> < Leq.lqz x : Leq Nat.zero x >
term  Leq.lqs : .[x : Nat] -> .[y : Nat] -> ^(y2 : Leq x y) -> < Leq.lqs x y y2 : Leq (Nat.succ x) (Nat.succ y) >
type  Increasing : - Size -> ^ Stream # -> Set
term  Increasing.inc : .[i : Size] -> .[x : Nat] -> .[y : Nat] -> ^(y3 : Leq x y) -> .[tl : Stream #] -> ^(y5 : Increasing i (Stream.cons [#] y tl)) -> < Increasing.inc i x y y3 tl y5 : Increasing $i (Stream.cons [#] x (Stream.cons [#] y tl)) >
type  Eq : ++(A : Set) -> ^(a : A) -> ^ A -> Set
term  Eq.refl : .[A : Set] -> .[a : A] -> < Eq.refl : Eq A a a >
term  proof : Eq (Stream #) (tail fib) (tail fib)
term  proof = Eq.refl
term  double : Stream # -> Stream #
term  double = \ s -> Stream.cons [?3] (head s) s
type  Bool : Set
term  Bool.tt : < Bool.tt : Bool >
term  Bool.ff : < Bool.ff : Bool >
term  leq : Nat -> Nat -> Bool
{ leq Nat.zero y = Bool.tt
; leq (Nat.succ x) Nat.zero = Bool.ff
; leq (Nat.succ x) (Nat.succ y) = leq x y
}
term  ite : Bool -> .[A : Set] -> A -> A -> A
{ ite Bool.tt [A] a1 a2 = a1
; ite Bool.ff [A] a1 a2 = a2
}
term  merge : .[i : Size] -> (Nat -> Nat -> Bool) -> Stream # -> Stream # -> Stream i
{ merge $[i < #] le (Stream.cons [.#] x xs) (Stream.cons [.#] y ys) = ite (le x y) [Stream $i] (Stream.cons [i] x (merge [i] le xs (Stream.cons [#] y ys))) (Stream.cons [i] y (merge [i] le (Stream.cons [#] x xs) ys))
}
term  first : .[A : Set] -> .[B : Set] -> A -> B -> A
{ first [A] [B] a b = a
}
term  map : .[i : Size] -> (Nat -> Nat) -> Stream i -> Stream i
{ map $[i < #] f (Stream.cons [.i] x xl) = Stream.cons [i] (f x) (map [i] f xl)
}
term  evil : .[i : Size] -> Stream i
{ evil $[i < #] = map [$i] (\ y -> Nat.succ y) (Stream.cons [i] Nat.zero (evil [i]))
}
--- evaluating ---
fibfour has whnf (head (tail (tail (tail (tail (fib' one one ?2))))))
fibfour evaluates to head (tail (tail (tail (tail (fib' (Nat.succ Nat.zero) (Nat.succ Nat.zero) ?2)))))
--- closing "fib.ma" ---
